rs/boundary_node/ic_boundary/src/salt_fetcher.rs

```diff
@@ -0,0 +1,137 @@
+use crate::core::Run;
+use anyhow::Error;
+use arc_swap::ArcSwapOption;
+use async_trait::async_trait;
+use candid::Principal;
+use candid::{Decode, Encode};
+use ic_canister_client::Agent;
+use ic_types::CanisterId;
+use prometheus::{
+    register_int_counter_vec_with_registry, register_int_gauge_with_registry, IntCounterVec,
+    IntGauge, Registry,
+};
+use salt_sharing_api::GetSaltResponse;
+use std::time::{SystemTime, UNIX_EPOCH};
+use std::{sync::Arc, time::Duration};
+use tokio::time::sleep;
+use tracing::warn;
+
+struct Metrics {
+    last_successful_fetch: IntGauge,
+    fetches: IntCounterVec,
+    last_salt_id: IntGauge,
+}
+
+impl Metrics {
+    fn new(registry: &Registry) -> Self {
+        Self {
+            last_successful_fetch: register_int_gauge_with_registry!(
+                format!("salt_sharing_last_successful_fetch"),
+                format!("The Unix timestamp of the last successful salt fetch"),
+                registry
+            )
+            .unwrap(),
+
+            last_salt_id: register_int_gauge_with_registry!(
+                format!("salt_sharing_last_salt_id"),
+                format!("ID of the latest fetched salt"),
+                registry,
+            )
+            .unwrap(),
+
+            fetches: register_int_counter_vec_with_registry!(
+                format!("salt_sharing_fetches"),
+                format!("Count of salt fetches and their outcome"),
+                &["result"],
+                registry
+            )
+            .unwrap(),
+        }
+    }
+}
+
+pub struct SharedSaltFetcher {
+    agent: Agent,
+    canister_id: CanisterId,
+    polling_interval: Duration,
+    anonymization_salt: Arc<ArcSwapOption<Vec<u8>>>,
+    metrics: Metrics,
+}
+
+impl SharedSaltFetcher {
+    pub fn new(
+        agent: Agent,
+        canister_id: Principal,
+        polling_interval: Duration,
+        anonymization_salt: Arc<ArcSwapOption<Vec<u8>>>,
+        registry: &Registry,
+    ) -> Self {
+        Self {
+            agent,
+            canister_id: CanisterId::try_from_principal_id(canister_id.into()).unwrap(),
+            anonymization_salt,
+            polling_interval,
+            metrics: Metrics::new(registry),
+        }
+    }
+}
+
+#[async_trait]
+impl Run for Arc<SharedSaltFetcher> {
+    async fn run(&mut self) -> Result<(), Error> {
+        loop {
+            let query_response = match self
+                .agent
+                .execute_query(&self.canister_id, "get_salt", Encode!().unwrap())
```

Comment:
As long as the data is not certified, we should use an update call for that.
```suggestion
                .execute_query(&self.canister_id, &self.canister_id, "get_salt", Encode!().unwrap(), nonce())
```
See [here](https://github.com/dfinity/ic/blob/master/rs/boundary_node/ic_boundary/src/rate_limiting/fetcher.rs#L73-L87)